code/datums/components/personalshield.dm

```diff
@@ -0,0 +1,148 @@
+	var/toggle_state = 1
+
+/datum/component/personalshield/Initialize(max_charges = 3, current_charges, recharge_delay = 20 SECONDS, recharge_speed = 2 SECONDS, recharge_rate = 1, must_be_worn_or_held = FALSE, shield_on = "shield-old", list/item_rechargable, list/togglable_states)
+	if(!istype(parent, /obj/item))
+		return COMPONENT_INCOMPATIBLE
+	src.max_charges = max_charges
+	if(!current_charges)
+		src.current_charges = max_charges
+	else
+		src.current_charges = current_charges
+	src.recharge_delay = recharge_delay
+	src.recharge_speed = recharge_speed
+	src.recharge_rate = recharge_rate
+	src.must_be_worn_or_held = must_be_worn_or_held
+	src.shield_on = shield_on
+	if(item_rechargable)
+		src.item_rechargable = typecacheof(item_rechargable)
+	src.togglable_states = togglable_states
+
+/datum/component/personalshield/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/equipped)
+	RegisterSignal(parent, COMSIG_ITEM_AFTER_PICKUP, .proc/pickup)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/dropped)
+	RegisterSignal(parent, COMPONENT_SHIELD_TOGGLE_COLOR, .proc/toggle_shield_color)
+
+/datum/component/personalshield/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_EQUIPPED, 
+		COMSIG_ITEM_AFTER_PICKUP, 
+		COMSIG_ITEM_DROPPED,
+		COMPONENT_SHIELD_TOGGLE_COLOR))
+
+/datum/component/personalshield/proc/toggle_shield_color(datum/source, mob/user)
+	if(wearer != user)
+		to_chat(user, "<span class='warning'>You can't interface with the shields's software if you haven't equipped it!</span>")
+		return
+	if(!length(togglable_states) || !current_charges)
+		to_chat(user, "<span class='warning'>You can't interface with the shields's software if the shield's broken!</span>")
+		return
+	to_chat(user, "<span class='warning'>You update the shield's hardware, changing back the shield's color.</span>")
+	if(++toggle_state > length(togglable_states))
+		toggle_state = 1
+	shield_on = togglable_states[toggle_state]
+	update_overlay(user)
+
+/datum/component/personalshield/proc/update_overlay(mob/user, state = "on")
+	if(!iscarbon(user))
+		return
+
+	wearer = user
+	wearer.remove_overlay(SHIELD_LAYER)
+	if(state == "off")
+		return
+	var/iconstate = "broken"
+	if(state == "on")
+		iconstate = shield_on
+	wearer.overlays_standing[SHIELD_LAYER] = list(mutable_appearance('icons/effects/effects.dmi', iconstate, MOB_LAYER + 0.01))
+	wearer.apply_overlay(SHIELD_LAYER)
+
+/datum/component/personalshield/proc/recharge()
+	current_charges = CLAMP((current_charges + recharge_rate), 0, max_charges)
+	if(wearer)
+		update_overlay(wearer)
+	if(current_charges == max_charges)
+		playsound(parent, 'sound/machines/ding.ogg', 50, TRUE)
+		return
+	playsound(parent, 'sound/magic/charge.ogg', 50, TRUE)
+	recharge_timer = addtimer(CALLBACK(src, .proc/recharge), recharge_speed, TIMER_STOPPABLE)
+
+/datum/component/personalshield/proc/start_recharge()
+	if(recharge_timer)
+		deltimer(recharge_timer)

```

Comment:
``recharge_timer = null`` after this.

Reply:
You don't need to null primitives

Reply:
Sure you don't need to but calling deltimer with null as an argument just seems odd.